### PR TITLE
V2 search experiments

### PR DESCRIPTION
**Pull request by whytheplatypus**

### CODE REVIEW

File: regcore/migrations/0012_migrate_documents.py [hunk 1]
```diff
@@ -176,6 +176,6 @@ class Migration(migrations.Migration):
     ]
 
     operations = [
-        migrations.RunPython(copy_regulations, uncopy_regulations),
-        migrations.RunPython(copy_preambles, uncopy_preambles),
+        #migrations.RunPython(copy_regulations, uncopy_regulations),
+        #migrations.RunPython(copy_preambles, uncopy_preambles),
```
Comment (whytheplatypus) [inline regcore/migrations/0012_migrate_documents.py]: shouldn't be committed.

File: regcore/models.py [hunk 2]
```diff
@@ -17,6 +22,59 @@ class Part(models.Model):
     class Meta:
         unique_together = ['name', 'title', 'date']
 
+    def find_in_structure(self, label):
+        return _walk_structure(self.structure, label)
+
+    def find_in_document(self, label):
+        return _walk_structure(self.document, label)
+    
+def _walk_structure(struct, label):
```
Comment (whytheplatypus) [inline regcore/models.py]: unused atm, remove

File: regcore/models.py [hunk 3]
```diff
@@ -17,6 +22,59 @@ class Part(models.Model):
...
+    if struct.get('label') == label:
+        return struct
+    for part in struct.get('children', []) or []:
+        result = _walk_structure(part, label)
+        if result:
+            return result
+            
+
+class SearchIndex(models.Model):
+    type = models.CharField(max_length=30)
```
Comment (whytheplatypus) [inline regcore/models.py]: this is an arbitrary max length, should be something more standard.

File: regcore/views.py [hunk 4]
```diff
@@ -84,3 +85,34 @@ def get_queryset(self):
...
+        fields = ("type", "content", "headline", "label", "parent", "regulation_title", "date")
+
+
+class SearchView(generics.ListAPIView):
+    serializer_class = SearchViewSerializer
+
+    def get_queryset(self):
+        q = self.request.query_params.get("q")
+        return SearchIndex.objects\
+            .filter(part__in=models.Subquery(Part.objects.order_by("name", "-date").distinct("name").values("id")))\
```
Comment (whytheplatypus) [inline regcore/views.py]: Move to a method on the object manager `current` and check for `<= today`

**Review by whytheplatypus — COMMENTED** [review]
Move this into a separate "app" that depends on regcore, allowing people to choose if they want to have this kind of search.

**Review by cgodwin1 — APPROVED** [review]

**Pull request merged**